adds part 1-3 of tutorial (#421)
temporalio/temporal-learning: docs/tutorials/ai/temporal-agentic-loop/primer/index.md

---
id: foundations-durable-ai
sidebar_position: 1
keywords: [ai, durable, temporal, llm, genai, agents, workflow]
tags: [AI, Series]
last_update:
  date: 2025-10-15
  author: Temporal Team
title: "Part 1: Foundations of Durable AI with a Research Application"
description: "Learn the fundamentals of building durable GenAI applications with Temporal by creating a research-to-PDF application"
image: /img/temporal-logo-twitter-card.png
---

# Part 1: Foundations of Durable AI with a Research Application

This tutorial introduces you to the importance of durability in GenAI applications. In the next tutorial, you'll learn how to use Temporal to add durability to your applications and handle failures gracefully.

By now, you've probably experienced generative AI firsthand. You've probably used ChatGPT and seen what LLMs can do. They excel at tasks like research, but their real power emerges when we connect them with users and other actions to build more advanced applications that go beyond simple chat interfaces.

> In this series, we'll build toward creating AI agents, but let's start with a simple chain:

Reviewer: ```suggestion
This series will guide you step by step toward creating AI agents. To begin, you'll start with a simple chain: using an LLM to generate research, then turning that research into a PDF.
```

Use an LLM to generate research ➝ then produce a PDF from that research.

### What are GenAI Applications?

> At their core, GenAI applications use an LLM as one component among many. The LLM isn't the application itself. Take ChatGPT as an example - it's an application that wraps an LLM, not an LLM itself. Even this seemingly simple chat interface does much more than just call an LLM:

Reviewer: ```suggestion
At their core, GenAI applications use an LLM as one component among many. The LLM isn't the application itself. Take ChatGPT as an example - it's an application that wraps an LLM, not an LLM itself. Even this seemingly simple chat interface does much more than just call an LLM, it:
```

- Displays responses to the user
- Captures user input
- Maintains conversation history
- Orchestrates each subsequent LLM call

Applications can look like many different formats. We will start with something like a chain workflow where a series of LLM calls, actions, and user interactions are strung together:

![Chain Workflow](https://images.ctfassets.net/0uuz8ydxyd9p/70SBemKQHnqfLxoHgPovQX/33f3a0b6cfc96eae2d17d1a463079560/Screenshot_2025-07-08_at_10.26.26%C3%A2__AM.png)

Applications can also be agentic, where the path through the business logic isn't predetermined. AI agents are GenAI applications where the LLM has agency over the functionality and flow of the application.

<img src="https://i.postimg.cc/zv5W9VkH/Screenshot-2025-10-07-at-7-48-20-PM.png" alt="Agentic Flow" width="40%" />

We'll first look at applications that look like a chain workflow.

### Prerequisites

You need an [Open AI API key](https://platform.openai.com/api-keys) for this tutorial.

### Setting Up Your Environment

> In this tutorial, you'll create a Research application that makes a call to the OpenAI API, conducts research on a topic of your choice, and generates a PDF report from that research. Let's start by setting up your environment.

Reviewer: ```suggestion
In this tutorial, you'll create a research application that makes a call to the OpenAI API, conducts research on a topic of your choice, and generates a PDF report from that research. Let's start by setting up your environment.
```

First, create your project:

1. Install [`uv`](https://docs.astral.sh/uv/), a Python package manager, on your machine:
```bash
pip install uv
```

2. Create your project: `uv init temporal-agentic-loop-tutorial`
3. Open your `temporal-agentic-loop-tutorial` project.
4. Install the packages you will need for this tutorial: `uv add python-dotenv litellm`.

### Create an `.env` File

Next you'll create a `.env` file to store your API keys. Create a new file named `.env` in your project directory and add the following:

```
LLM_API_KEY=YOUR_API_KEY
LLM_MODEL=openai/gpt-4o
```

> By default this tutorial uses OpenAI's GPT-4o. If you want to use a different LLM provider, look up the appropriate model name [in the LiteLLM documentation](https://docs.litellm.ai/docs/providers) and change the `LLM_MODEL` field and provide your API key.

Reviewer: ```suggestion
By default, this tutorial uses OpenAI's GPT-4o. If you want to use a different LLM provider, look up the appropriate model name [in the LiteLLM documentation](https://docs.litellm.ai/docs/providers) and change the `LLM_MODEL` field and provide your API key.
```

    - Sends your prompt to the specified LLM model 
    - Includes the API key for authentication
    - Formats the prompt as a message with role "user" (as LLMs expect a conversation format)

```python
import os
from dotenv import load_dotenv
from litellm import completion
from litellm.types.utils import ModelResponse
from models import LLMCallInput

load_dotenv(override=True) # Reads your .env file and loads your environment variables

# Get LLM_API_KEY environment variable
LLM_MODEL = os.getenv("LLM_MODEL", "openai/gpt-4o")
LLM_API_KEY = os.getenv("LLM_API_KEY", None)

def llm_call(input: LLMCallInput) -> ModelResponse:
    return completion(
        model=input.llm_model,
        api_key=input.llm_api_key,
        messages=[{"content": input.prompt, "role": "user"}],
    )
```

Let's test this with a simple prompt. Add the following code to `app.py`:

```python
prompt = "Give me 5 facts about elephants."
llm_input = LLMCallInput(prompt=prompt, llm_api_key=LLM_API_KEY, llm_model=LLM_MODEL)
result = llm_call(llm_input)

# Extract the text content
content = result.choices[0].message.content
print(content)
```

Now run this file with `uv run app.py`. 

> You should see your output from your LLM in your terminal window!

Reviewer: ```suggestion
You should see your output from your LLM in your terminal window.
```
